### ISSUE TITLE
Question about the exact LSD-HF evaluation protocol

### ISSUE THREAD

**Issue by woaikanxue** [woongzip1/UniverSR#6]
Hi UniverSR authors,

Thank you for releasing the paper and the official implementation. I am currently trying to reproduce and compare the LSD-HF results reported in your paper, especially the `L` metric in Table 2.

I found the official metric script:

`scripts/compute_metrics.py`

Based on the code, my understanding is that LSD is computed as follows:

1. Both the generated audio and the reference audio are loaded.
2. If the sampling rate is not 48 kHz, the audio is resampled to 48 kHz.
3. Multi-channel audio is converted to mono by averaging channels.
4. The generated and reference signals are truncated to the same length.
5. STFT magnitude is computed with:
   - `n_fft = 1024`
   - `hop_length = 256`
   - Hann window
   - `center=True`, if using the PyTorch default
6. The log spectrum is computed as:

    ```python
    log_spec = torch.log10(abs(STFT(x)) ** 2 clipped at 1e-6)
    ```

    corresponding to the code:

    ```python
    torch.log10(stft_magnitude(x).square().clamp(min=1e-6))
    ```

7. LSD is computed as the root mean square difference over frequency bins, followed by averaging over time frames and then averaging over files:

    ```python
    lsd = sqrt(mean_freq((log_spec_pred - log_spec_ref) ** 2)).mean_time()
    ```

8. For LSD-HF, the frequency bins above the cutoff are used:

    ```python
    SR_TO_CUTOFF_BIN = {
        8: 80,
        12: 128,
        16: 170,
        24: 256,
    }
    ```

I would like to confirm whether this is exactly the evaluation protocol used to report the LSD-HF values in Table 2.

Specifically, I have the following questions:

1. Is the `L` metric in Table 2 exactly the `lsd_high` value produced by `scripts/compute_metrics.py`?

2. Is the log spectrum intentionally computed as:

    ```python
    log10(|STFT(x)|^2)
    ```

    without multiplying by 10?

    In many LSD implementations, the log spectrum is computed in dB scale, for example:

    ```python
    10 * log10(|STFT(x)|^2)
    ```

    or:

    ```python
    20 * log10(|STFT(x)|)
    ```

    However, the released script appears to use only:

    ```python
    log10(|STFT(x)|^2)
    ```

    Could you confirm that this scale is the one used for the reported paper results?

3. Were the Table 2 LSD-HF values computed with the following STFT parameters?

    ```text
    target sampling rate = 48000 Hz
    n_fft = 1024
    hop_length = 256
    window = Hann(1024)
    center = True
    ```

4. For LSD-HF, were the cutoff bins exactly the following?

    ```python
    {8k: 80, 12k: 128, 16k: 170, 24k: 256}
    ```

    I noticed that for 8 kHz input, bin 80 at 48 kHz / 1024 corresponds to approximately 3750 Hz rather than exactly 4000 Hz. Is this cutoff chosen to match the model configuration rather than the exact Nyquist frequency?

5. How is the final dataset-level LSD-HF averaged?

    Is it computed as:

    ```text
    compute utterance-level LSD-HF first
    then average over all utterances
    ```

    or is it pooled over all frames from all utterances before averaging?

6. Was any additional preprocessing applied before computing LSD-HF, such as:
   - silence trimming,
   - loudness normalization,
   - peak normalization,
   - waveform alignment,
   - padding removal,
   - or any filtering?

7. What exact VCTK test split was used to compute the LSD-HF values in Table 2?

    Could you share the speaker IDs and the number of utterances used for the LSD-HF evaluation?

8. For the FlowHigh baseline in Table 2, were the FlowHigh outputs generated by your team using the public FlowHigh checkpoint/code, or were the LSD-HF values taken from the original FlowHigh paper?

    If they were generated by your team, could you share the checkpoint/inference settings used before computing LSD-HF?

I am asking because when I run the released `scripts/compute_metrics.py` on my generated 48 kHz outputs and my held-out VCTK test set, I obtain LSD-HF values on a different scale from the values reported in Table 2. I would like to make sure that my evaluation protocol is aligned with yours before making any comparison.

Thank you very much for your help and for open-sourcing the project.

**Comment by woongzip1** [comment]
Hi, thank you for your interest in our work and for the detailed investigation. 
To address your questions:

- **Metric Identification**: Yes, the $L$ metric in Table 2 is exactly the `lsd_high` value produced by `scripts/compute_metrics.py`.
- **Log Spectrum Scale**: We confirm that the log spectrum was computed as $\log_{10}(|\text{STFT}(x)|^2)$ without the $10$ or $20$ multiplier. Our implementation is based on the [AP-BWE](https://github.com/yxlu-0102/AP-BWE/blob/main/cal_metrics.py) repository, extended to calculate LSD in the generated HF region.
- **STFT Parameters**: Yes, the parameters you listed are exactly what we used for the paper.
- **Cutoff Bins**: The bins were chosen to match our model's configuration. Bin 80 for 8 kHz was intended to mitigate overlap issues during HF concatenation, as previously discussed in our issues section.
- **Averaging Method**: We followed utterance-level averaging. We first compute the LSD-HF for each utterance and then calculate the arithmetic mean across the test set.
- **Preprocessing**: For a fair evaluation, we only applied standard preprocessing: silence trimming and waveform alignment. We compared the metrics by directly evaluating the files between the input and generated directories without any additional processing like loudness normalization.
- **VCTK Test Split**: We used speakers **p280** and **s5** as mentioned in the paper.
- **FlowHigh Baseline**: We evaluated FlowHigh using the official checkpoints and code provided by the authors. You can find their implementation [here](https://github.com/jjunak-yun/FLowHigh_code).

We hope this clarifies the discrepancies. 
Best regards.
